ThingIFSDK/ThingIFSDK/ThingIFAPI+Persistence.swift

```diff
@@ -0,0 +1,180 @@
+    {
+        let baseKey = ThingIFAPI.SHARED_NSUSERDEFAULT_KEY_INSTANCE
+        let versionKey = ThingIFAPI.getStoredSDKVersionKey(tag)
+        let key = ThingIFAPI.getStoredInstanceKey(tag)
+
+        // try to get iotAPI from NSUserDefaults
+        guard let dict =
+                UserDefaults.standard.dictionary(forKey: baseKey) else {
+            throw ThingIFError.apiNotStored(tag: tag)
+        }
+
+        if dict[key] == nil  {
+            throw ThingIFError.apiNotStored(tag: tag)
+        }
+
+        let storedSDKVersion = dict[versionKey] as? String
+        if !isLoadable(storedSDKVersion) {
+            throw ThingIFError.apiUnloadable(
+              tag: tag,
+              storedVersion: storedSDKVersion,
+              minimumVersion: MINIMUM_LOADABLE_SDK_VERSION)
+        }
+
+        guard let data = dict[key] as? Data else {
+            throw ThingIFError.invalidStoredApi
+        }
+        guard let decoder = Decoder(data) else {
+            throw ThingIFError.invalidStoredApi
+        }
+        guard let retval = ThingIFAPI.deserialize(decoder) as? ThingIFAPI else {
+            throw ThingIFError.invalidStoredApi
+        }
+        return retval
+
+    }
+    /** Save this instance
+    */
+    open func saveInstance() -> Void {
+        self.saveToUserDefault()
+    }
+
+    /** Clear all saved instances in the NSUserDefaults.
+    */
+    open static func removeAllStoredInstances() {
+        let baseKey = ThingIFAPI.SHARED_NSUSERDEFAULT_KEY_INSTANCE
+        UserDefaults.standard.removeObject(forKey: baseKey)
+        UserDefaults.standard.synchronize()
+    }
+
+    /** Remove saved specified instance in the NSUserDefaults.
+    - Parameter tag: tag of the ThingIFAPI instance or nil for default tag
+    */
+    open static func removeStoredInstances(_ tag : String? = nil) -> Void {
+        let baseKey = ThingIFAPI.SHARED_NSUSERDEFAULT_KEY_INSTANCE
+        let versionKey = ThingIFAPI.getStoredSDKVersionKey(tag)
+        let key = ThingIFAPI.getStoredInstanceKey(tag)
+        if var dict = UserDefaults.standard.dictionary(forKey: baseKey) {
+            dict[versionKey] = nil
+            dict[key] = nil
+            UserDefaults.standard.set(dict, forKey: baseKey)
+            UserDefaults.standard.synchronize()
+        }
+    }
+
+    internal func saveToUserDefault() -> Void {
+        let baseKey = ThingIFAPI.SHARED_NSUSERDEFAULT_KEY_INSTANCE
+        let versionKey = ThingIFAPI.getStoredSDKVersionKey(self.tag)
+        let key = ThingIFAPI.getStoredInstanceKey(self.tag)
+
+        var coder = Coder()
+        serialize(&coder)
+        let data = coder.finishCoding()
+
+        if var dict = UserDefaults.standard.dictionary(forKey: baseKey) {
+            dict[versionKey] = SDKVersion.sharedInstance.versionString
+            dict[key] = data
+            UserDefaults.standard.set(dict, forKey: baseKey)
+        } else {
+            UserDefaults.standard.set([key:data], forKey: baseKey)
```

Comment:
should `[key:data]` be `[key:data,  versionKey: SDKVersion.sharedInstance.versionString ]`?